streaming/python/function.py

```diff
@@ -0,0 +1,315 @@
+import importlib
+import inspect
+import sys
+from abc import ABC, abstractmethod
+import typing
+
+import cloudpickle
+from ray.streaming.runtime import gateway_client
+
+
+class Function(ABC):
+    """The base interface for all user-defined functions."""
+
+    def open(self, conf: typing.Dict[str, str]):
+        pass
+
+    def close(self):
+        pass
+
+
+class SourceContext(ABC):
+    """
+    Interface that source functions use to emit elements, and possibly
+     watermarks."""
+
+    @abstractmethod
+    def collect(self, element):
+        """Emits one element from the source, without attaching a timestamp."""
+        pass
+
+
+class SourceFunction(Function):
+    """Interface of Source functions."""
+
+    @abstractmethod
+    def init(self, parallel, index):
+        """
+        Args:
+            parallel: parallelism of source function
+            index: task index of this function and goes up from 0 to
+             parallel-1.
+        """
+        pass
+
+    @abstractmethod
+    def run(self, ctx: SourceContext):
+        """Starts the source. Implementations can use the
+         :class:`SourceContext` to emit elements.
+        """
+        pass
+
+    def close(self):
+        pass
+
+
+class MapFunction(Function):
+    """
+    Base interface for Map functions. Map functions take elements and transform
+    them element wise. A Map function always produces a single result element
+    for each input element.
+    """
+
+    def map(self, value):
+        pass
+
+
+class FlatMapFunction(Function):
+    """
+    Base interface for flatMap functions. FlatMap functions take elements and
+    transform them into zero, one, or more elements.
+    """
+
+    def flat_map(self, value, collector):
+        """Takes an element from the input data set and transforms it into zero,
+        one, or more elements.
+
+        Args:
+            value: The input value.
+            collector: The collector for returning result values.
+        """
+        pass
+
+
+class FilterFunction(Function):
+    """
+    A filter function is a predicate applied individually to each record.
+    The predicate decides whether to keep the element, or to discard it.
+    """
+
+    def filter(self, value):
+        """The filter function that evaluates the predicate.
+
+        Args:
+            value: The value to be filtered.
+
+        Returns:
+            True for values that should be retained, false for values to be
+            filtered out.
+        """
+        pass
+
+
+class KeyFunction(Function):
```

Comment:
Why call this function `KeyFunction` with `key_by` method？


Reply:
It's a `KeyFunction ` used to extract the key. I think it's ok to name it as `KeyFunction`